text/0000-explicit-tail-calls.md

```diff
@@ -0,0 +1,723 @@
+
+This feature will have interactions with other features that depend on stack frames, for example, debugging and backtraces. See [drawbacks](#drawbacks) for further discussion.
+
+See below for specifics on interations with other features.
+
+## Coercions of the Tail Called Function's Return Type
+
+All coercions that do any work (like deref coercion, unsize coercion, etc) are prohibited.
+Lifetime-shortening coercions (`&'static T` -> `&'a T`) are allowed but will be checked by the borrow checker.
+
+Note that, while in theory, never-to-any coercions (`! -> T`) could be allowed, they are difficult to implement and require backend support. As a result they are not allowed as per this RFC. This has no effect on using macros like `panic!()` as they are not functions, affected are only functions like the following:
+
+```rust
+fn never() -> ! {
+    loop {}
+}
+
+fn tail_call_never_type() -> usize {
+    become never();
+}
+```
+
+## Closures
+[closures]: #closures
+
+Tail calling closures _and_ tail calling _from_ closures is **not** allowed.
+This is due to the high implementation effort, see below, this restriction can be lifted by a future RFC.
+
+Closures use the `rust-call` unstable calling convention, which would need to be adapted to guarantee TCE.
+Additionally, any closure that has captures would need special handling, since the captures would currently be dropped before the tail call.
+
+## Variadic functions using `c_variadic`
+
+Tail calling [variadic functions](https://doc.rust-lang.org/beta/unstable-book/language-features/c-variadic.html) _and_ tail calling _from_ variadic functions is **not** allowed.
+As support for variadic function is stabilized on a per target level, support for tail-calls regarding variadic functions would need to follow a similar approach. To avoid this complexity and to minimize implementation effort for backends, this interaction is currently not allowed but supported can be added with a future RFC.
+
+## Generators
+
+Tail calling [generators](https://doc.rust-lang.org/beta/unstable-book/language-features/generators.html) is **not** allowed as it is a fundamental mismatch of functionality. Generators expect to be called multiple times `yield`ing values, however, when using a tail call control would never be returned to the calling function.
+
+Tail calling from generators is also **not** allowed, as the generator state is stored internally, tail calling from the generator function would require additional support to function correctly. To limit implementation effort this is not supported but can be supported by a future RFC.
+
+## Async
+
+Tail calling async functions is **not** allowed as it requires special support for the async state machine.
```

Comment:
updated here: https://github.com/rust-lang/rfcs/pull/3407/commits/b3dc340150cd6244baa91eff0220e6b66dcfee3c

Please give it another read, I'm not that familiar with async.

Reply:
Shouldn't `become f().await` be the same as `become (f().await)`?  I.e. it should mean:

```rust
let _tmp = f().await;
become _tmp;
```

So if `f().await` returns a compatible function pointer then this would be legal.

Reply:
I've written a few comments on the async section in the review down below.